utils/cmd/cmd.c

```diff
@@ -0,0 +1,721 @@
+		print_error("Empty response message");
+		goto fail;
+	}
+
+	DBusMessageIter iter_codecs;
+	for (dbus_message_iter_recurse(&iter, &iter_codecs);
+			dbus_message_iter_get_arg_type(&iter_codecs) != DBUS_TYPE_INVALID;
+			dbus_message_iter_next(&iter_codecs)) {
+
+		if (dbus_message_iter_get_arg_type(&iter_codecs) != DBUS_TYPE_DICT_ENTRY)  {
+			print_error("item is not dict entry");
+			goto fail;
+		}
+
+		DBusMessageIter iter_codecs_entry;
+		dbus_message_iter_recurse(&iter_codecs, &iter_codecs_entry);
+
+		if (dbus_message_iter_get_arg_type(&iter_codecs_entry) != DBUS_TYPE_STRING) {
+			print_error("item is not string");
+			goto fail;
+		}
+
+		const char *codec;
+		dbus_message_iter_get_basic(&iter_codecs_entry, &codec);
+		printf("%s\n", codec);
+
+		/* Ignore the properties field, get next codec. */
+	}
+	result = EXIT_SUCCESS;
+
+fail:
+	dbus_message_unref(rep);
+	dbus_message_unref(msg);
+
+	return result;
+}
+
+static int select_codec(int argc, char *argv[]) {
+	if (argc != 3) {
+		print_error_usage("Invalid arguments.");
+		return EXIT_FAILURE;
+	}
+
+	const char *path = argv[1];
+	const char *codec = argv[2];
+
+	if (!check_path(path)) {
+		return EXIT_FAILURE;
+	}
+
+	DBusMessage *msg;
+	DBusMessage *rep;
+	DBusError err = DBUS_ERROR_INIT;
+	if ((msg = dbus_message_new_method_call(dbus_ctx.ba_service, path,
+					BLUEALSA_INTERFACE_PCM, "SelectCodec")) == NULL) {
+		print_error("Out of memory");
+		return EXIT_FAILURE;
+	}
+
+	DBusMessageIter iter;
+	dbus_message_iter_init_append(msg, &iter);
+
+	if (!dbus_message_iter_append_basic(&iter, DBUS_TYPE_STRING, &codec)) {
+		print_error("Couldn't select BlueALSA PCM Codec: %s", err.message);
+		dbus_message_unref(msg);
+		return EXIT_FAILURE;
+	}
+
+	DBusMessageIter props = DBUS_MESSAGE_ITER_INIT_CLOSED;
+	if (!dbus_message_iter_open_container(&iter, DBUS_TYPE_ARRAY, "{sv}", &props)) {
+		print_error("Couldn't select BlueALSA PCM Codec: %s", err.message);
+		dbus_message_unref(msg);
+		return EXIT_FAILURE;
+	}
+	dbus_message_iter_close_container(&iter, &props);
+
+	if ((rep = dbus_connection_send_with_reply_and_block(dbus_ctx.conn,
+					msg, DBUS_TIMEOUT_USE_DEFAULT, &err)) == NULL) {
+		dbus_message_unref(msg);
+		print_error("Couldn't select BlueALSA PCM Codec: %s", err.message);
+		return EXIT_FAILURE;
+	}
+
+	dbus_message_unref(rep);
+	dbus_message_unref(msg);
+	return EXIT_SUCCESS;
+}
+
+static int properties(int argc, char *argv[]) {
+	if (argc != 2) {
+		print_error_usage("Too many arguments.");
+		return EXIT_FAILURE;
+	}
+
+	const char *path = argv[1];
+
+	struct ba_pcm pcm;
+	if (!get_pcm(path, &pcm)) {
+		print_error("Invalid pcm path: %s", path);
+		return EXIT_FAILURE;
+	}
+
+	/* Get the transport and mode values from the pcm path */
+	char *token[6] = { 0 };
+	char *tmp = strdup(path);
+	char *next = tmp + 1;
+	size_t i = 0;
+	while (next && i < 6) {
+		token[i++] = strsep(&next, "/");
+	}
+
+	printf("Device: %s\n", pcm.device_path);
+
+	if (strstr(token[4], "a2dpsrc"))
```

Comment:
Maybe it will be worth adding transport type string to the pcm struct (or extending existing `profile` variable) in the dbus-client.h ?

Reply:
You can base this PR on this branch: https://github.com/Arkq/bluez-alsa/tree/update-dbus-client I've updated the internal client code to match with the current BlueALSA D-Bus API.

Reply:
see commit 1d92201